### PR TITLE
test(): add playwright for UI testing

### PR DESCRIPTION
**Pull request by luciemars**
## Motivation

Add Playwright UI testing framework.

## Description

Add playwright dependency with config file and examples of test (1 passing and 1 failing).
Running the tests will open the test report in a browser.

### CODE REVIEW

File: e2e/tests/example.spec.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { test, expect } from '@playwright/test';
+
+test('has title', async ({ page }) => {
+  await page.goto('https://playwright.dev/');
```
Comment (asturur) [inline e2e/tests/example.spec.ts:4]: Hi @luciemars thanks for starting this!
Do you know if i can happily serve a local file here for my test setup or if i have to run a webserver?

Comment (luciemars) [reply e2e/tests/example.spec.ts:4]: I found this in the docs https://playwright.dev/docs/test-advanced#launching-a-development-web-server-during-the-tests.
I can look a bit more if you want to do it differently.

Comment (asturur) [reply e2e/tests/example.spec.ts:4]: no no i can look that myself no worries, i was just wondering if you had an answer.
I want to add a real fabricJS test before merging, i ll keep you posted

**Comment by ShaMan123** [comment]
Can you guys explain what tests do you want to run with this framework?

**Comment by ShaMan123** [comment]
Are we going to move from the current framework to this?

**Comment by asturur** [comment]
This could provide interaction tests we have difficulties emulating with qunit.
For sure not moving but just adding different kind of tests, like drag a specific corner, or drag an object or shift click interactions. I didn't test it yet but seems one that solve those issues

**Comment by asturur** [comment]
@luciemars @ShaMan123 slowly writing my first test.
just have a look at the files to keep updated on what i m trying

File: playwright.config.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        ...devices['Desktop Chrome'],
+      },
+    },
+  ],
+
+  /* Run your local dev server before starting the tests */
+  // webServer: {
+  //   command: 'npm run start',
+  //   port: 3000,
+  // },
```
Comment (asturur) [inline playwright.config.ts:61]: ohhh this is what i needed, i can remove concurrently.
Nice that playwright docs do not put this out clearly somewhere

File: playwright.config.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  use: {
+    /* Maximum time each action such as `click()` can take. Defaults to 0 (no limit). */
+    actionTimeout: 0,
+    /* Base URL to use in actions like `await page.goto('/')`. */
+    // baseURL: 'http://localhost:3000',
+
+    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
+    trace: 'retain-on-failure',
+    video: 'retain-on-failure',
+    screenshot: 'only-on-failure',
```
Comment (asturur) [inline playwright.config.ts:43]: @ShaMan123 playwright can retain a video or a screenshot of what went wrong in the test.
I think this could be very useful once we know how to make it work with github actions

File: e2e/tests/example.spec.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+  await page.evaluate(() => {
+    const textbox = new fabric.Textbox('HELLO');
+    fabricCanvas.add(textbox);
+  });
+
+  await page.click('canvas.upper-canvas', {
+    position: { x: 15, y: 15 },
+    clickCount: 2,
+    delay: 200
+  });
```
Comment (asturur) [inline e2e/tests/example.spec.ts:14]: we would probably need some helpers, like click on something. drag control X.
Becuase the playwright test recorder doesn't record the exact position where it clicks ( maybe we could see if is fixable )

**Comment by asturur** [comment]
its getting better.
The real issue here is figuring out how much scope a test should have.
For example a text could cover functionality of the default object control set. includin shift and alt combinations.
Probably that would be a good first test.

File: e2e/tests/text/text-editing.spec.ts [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  // click the object to select it
+  await clickCanvas(page, {
+    position: textCenter,
+  });
+
+  // drag the control
+  await page.mouse.move(controlPoint.x, controlPoint.y);
+  await page.mouse.down();
+  await page.mouse.move(controlPoint.x + 300, controlPoint.y, { steps: 40 });
+  await page.mouse.up();
```
Comment (ShaMan123) [inline e2e/tests/text/text-editing.spec.ts:52]: this is awesome

File: playwright.config.ts (hunk 3 above)
Comment (ShaMan123) [reply playwright.config.ts:43]: Wow
Thinking of a video of a control transforming an object 🤯
This can help us cover fabric and really test it w/o battling logic

File: README.md [hunk 6]
```diff
@@ -191,6 +191,14 @@ http
 
 </details>
 
+## UI tests
+
+Run UI tests with Playwright in parallel in Chrome (headed mode).
+
+```bash
+$ npm run test:e2e
```
Comment (ShaMan123) [inline README.md:199]: I want it to become part of the cli and not standalone
Something like:
`npm test -- -s e2e`

**Review by ShaMan123 — COMMENTED** [review]
Cool stuff

File: e2e/tests/text/text-editing.spec.ts (hunk 5 above)
Comment (ShaMan123) [reply e2e/tests/text/text-editing.spec.ts:52]: does it fire a single move? or repeats the move from down?

Comment (asturur) [reply e2e/tests/text/text-editing.spec.ts:52]: i didn't measure but you can totally see the text being resized smoothly so probably it fires the 40 steps i specified there.

File: README.md (hunk 6 above)
Comment (asturur) [reply README.md:199]: yes for now we just need to have it working with a starting test.
Then we need to understand how reliable those are, but definitely will end up in the CI.

**Comment by ShaMan123** [comment]
Did you run the build cmd?
npm rub build -- -f -w

**Comment by ShaMan123** [comment]
@luciemars how familiar are you with playwright?
Is there a way to test performance with it?

**Comment by ShaMan123** [comment]
I see that visual testing is possible out of the box
https://playwright.dev/docs/test-configuration#expect-options
https://playwright.dev/docs/api/class-snapshotassertions#snapshot-assertions-to-match-snapshot-1

**Comment by ShaMan123** [comment]
I don't understand how to import fabric as a module
I want these tests to use esm.

**Comment by ShaMan123** [comment]
test-finished-1
[video.webm](https://github.com/fabricjs/fabric.js/assets/34343793/738a3a47-097c-4d40-922a-ee45e1c79e69)

WOW

**Comment by asturur** [comment]
let's merge this and jest too.
So we can add slowly new types of tests and learn from there
